JENKINS-11888 Add stop-job CLI command

Austry opened the pull request:
See [JENKINS-11888](https://issues.jenkins-ci.org/browse/JENKINS-11888).

### Proposed changelog entries
* RFE: Added `stop-job` CLI command which allows aborting builds

### Submitter checklist
- [X] JIRA issue is well described
- [X] Changelog entry appropriate for the audience affected by the change
- [X] Autotests

Commit:
JENKINS-11888 Add stop-job CLI command

Commit:
JENKINS-11888 Add license headers

Austry commented on core/src/main/java/hudson/cli/StopJobCommand.java:
```diff
@@ -0,0 +1,109 @@
+package hudson.cli;
```
fixed

Austry commented on core/src/main/java/hudson/cli/ReloadJobCommand.java:
```diff
@@ -54,7 +54,6 @@
 
     @Override
     public String getShortDescription() {
-
```
fixed

Austry commented on core/src/main/java/hudson/cli/StopJobCommand.java:
```diff
@@ -0,0 +1,109 @@
+        names.addAll(jobNames);
+        final StringBuilder resultBuilder = new StringBuilder();
+        for (final String jobName : names) {
+            AbstractProject job = null;
```
I have not found a better way to find all running builds. Can you give some advice, please?

daniel-beck replied:
`Job`/`Run`? Or are you using API specific to these subtypes? I cannot see any.

Austry replied:
I am using `AbstractBuild#doStop` and can't find alternative in 'Job' or 'Run'

oleg-nenashev replied:
Yes, there is no method available on the `Run` level. @jglick @svanoort I believe we could introduce a method with a default noop implementation (or exception for non-supported op). Code in AbstractBuild and WorkflowRun is equivalent: https://github.com/jenkinsci/workflow-job-plugin/blob/master/src/main/java/org/jenkinsci/plugins/workflow/job/WorkflowRun.java#L822-L831 and https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/model/AbstractBuild.java#L1315-L1324 , but a failing defauld implementation may be useful for other Run types. WDYT?

oleg-nenashev commented on core/src/main/java/hudson/cli/StopJobCommand.java:
```diff
@@ -0,0 +1,132 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.cli;
```
I would rather recommend `jenkins.cli` as a package, but ok

daniel-beck replied:
Agree. New code should be in `jenkins.*`

oleg-nenashev commented on core/src/main/java/hudson/cli/StopJobCommand.java:
```diff
@@ -0,0 +1,132 @@
+            if (job != null) {
+                stopJobBuilds(job, jobName, resultBuilder);
+            } else {
+                resultBuilder.append(String.format("Job with name %s not found.\n", jobName));
```
Incorrect message. The job just may have wrong type

oleg-nenashev reviewed (commented):
I would also prefer to have a generic implementation for any `Run` type if possible (with error when operation is not supported). Otherwise we will quickly end up with bug reports taking the Pipeline adoption into account

Commit:
JENKINS-11888 Review fix

Austry commented:
I was forced to rename `jenkins/core/src/main/resources/jenkins/CLI` to `jenkins/core/src/main/resources/jenkins/cli` in order to move `StopJobCommand` to `jenkins.cli`
package
without what I constantly receive error in logs and `StopJobCommand` not injects to commands list

error trace
```
Oct 20, 2018 6:32:25 PM hudson.ExtensionFinder$Sezpoz _find
WARNING: Failed to load jenkins.cli.StopJobCommand
java.lang.InstantiationException: /Users/austry/.jenkins/war/WEB-INF/lib/jenkins-core-2.145-SNAPSHOT.jar might need to be rebuilt: java.lang.ClassNotFoundException: cached miss for jenkins.cli.StopJobCommand
	at net.java.sezpoz.IndexItem.element(IndexItem.java:144)
	at hudson.ExtensionFinder$Sezpoz._find(ExtensionFinder.java:701)
	at hudson.ExtensionFinder$Sezpoz.find(ExtensionFinder.java:690)
	at hudson.ClassicPluginStrategy.findComponents(ClassicPluginStrategy.java:494)
	at hudson.ExtensionList.load(ExtensionList.java:380)
	at hudson.ExtensionList.ensureLoaded(ExtensionList.java:318)
	at hudson.ExtensionList.getComponents(ExtensionList.java:183)
	at jenkins.model.Jenkins$6.onInitMilestoneAttained(Jenkins.java:1097)
	at jenkins.InitReactorRunner$1.onAttained(InitReactorRunner.java:83)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.lambda$onAttained$3(ReactorListener.java:102)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.run(ReactorListener.java:109)
	at org.jvnet.hudson.reactor.ReactorListener$Aggregator.onAttained(ReactorListener.java:102)
	at org.jvnet.hudson.reactor.Reactor$1.run(Reactor.java:177)
	at org.jvnet.hudson.reactor.Reactor$Node.run(Reactor.java:117)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: cached miss for jenkins.cli.StopJobCommand
	at hudson.PluginManager$UberClassLoader.findClass(PluginManager.java:1878)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at net.java.sezpoz.IndexItem.element(IndexItem.java:134)
	... 16 more
```

oleg-nenashev commented on core/src/main/java/jenkins/cli/StopJobCommand.java:
```diff
@@ -0,0 +1,130 @@
+import java.util.logging.Logger;
+
+@Extension
+public class StopJobCommand extends CLICommand {
```
It would be rather `StopBuildsCommand` IMHO

jglick replied:
Yes.

oleg-nenashev commented on core/src/main/java/jenkins/cli/StopJobCommand.java:
```diff
@@ -0,0 +1,130 @@
+            if (item instanceof AbstractProject) {
+                stopJobBuilds((AbstractProject) item, jobName, resultBuilder);
+            } else if (item != null) {
+                resultBuilder.append("Job have not supported type.\n");
```
The message needs some massaging

oleg-nenashev replied:
```suggestion
                resultBuilder.append(String.format("Cannot abort runs for %s. Unsupported job type: %s\n", item, item.getClass()));
```

jglick reviewed (changes requested):
Reminder that this needs to be rewritten to use `Run.getExecutor()`.

oleg-nenashev commented:
If @austry is not available, my plan is to actually merge it as a
Freestyle-only solution and to create a follow-up for Pipeline support.
Should be acceptable if it is explicit in the changelog

On Mon, Feb 4, 2019 at 6:28 PM Jesse Glick <notifications@github.com> wrote:

> *@jglick* requested changes on this pull request.
>
> Reminder that this needs to be rewritten to use Run.getExecutor().
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/3686#pullrequestreview-199709436>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AC3IoK4jhX-UF6b5mA5GEkWxt3O3uuESks5vKG3agaJpZM4XMC99>
> .
>

Austry commented:
I am available. please give me some time to fix review and try to implement @jglick suggestion

Commit:
JENKINS-11888 Fix review

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,135 @@
+            if (item instanceof Job) {
+                stopJobBuilds((Job) item, jobName, resultBuilder);
+            } else if (item != null) {
+                resultBuilder.append(String.format("Cannot abort runs for %s. Unsupported job type: %s\n", item, item.getClass()));
```
I would suggest checking all the arguments first, collecting a `List<Job>` to process, and throwing `IllegalArgumentException` if an argument is encountered which is not a valid `Job`.

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,135 @@
+        final Run lastBuild = job.getLastBuild();
+        final List<String> stoppedBuildsNames = new ArrayList<>();
+        if (lastBuild != null) {
+            stopBuild(lastBuild, jobName, stoppedBuildsNames);
```
Since it was not immediately obvious to me—this _does_ verify that this build is in progress, inside `stopBuild`. (Since jobs may be concurrent-capable, the default for Pipeline in fact, the last build may be complete even while preceding builds are not.) So then the call from `checkAndStopPreviousBuilds` is redundantly checking `isBuilding`. Probably harmless if a bit confusing. I would suggest moving the `isBuilding` check to this call site since it is only for `getLastBuild` that we do not know whether or not we are dealing with an in-progress build. (Anyway a completed build would simply return null from `getExecutor`.)

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,135 @@
+        if (stoppedBuildNames.isEmpty()) {
+            result.append(String.format("No builds stopped for job '%s'", jobName));
+        } else {
+            result.append(String.format("Builds stopped for job '%s': ", jobName));
```
The use of `StringBuilder` seems unnecessary in this class. You can simply print to `stdout` immediately when doing something. That would make the output more responsive, too, in case the command takes a while to run.

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,135 @@
+    }
+
+    private void logBuildStopped(final String jobName, final String buildName) {
+        LOGGER.log(Level.INFO, String.format("Build %s in job %s aborted", buildName, jobName));
```
Probably unnecessary. Jenkins already logs a message when a build terminates for any reason.

jglick commented on core/src/test/java/jenkins/cli/StopBuildsCommandTest.java:
```diff
@@ -0,0 +1,180 @@
+import static org.mockito.Mockito.verify;
+import static org.powermock.api.mockito.PowerMockito.*;
+
+@RunWith(PowerMockRunner.class)
```
I tend to not trust any tests written with PowerMock. They just prove what I could already tell by looking at the source code. They do not prove that the feature actually works at all. YMMV

Austry replied:
It's seems like a valid technique for mocking` Jenkins.get()` method. In other tests it works in the same way (`NewViewLinkTest` for example). If you insist to change it, please show me how to inject `Jenkins` instance in `StopBuildsCommand` without `PowerMock`.

jglick replied:
I know there are other tests using mocks; I do not trust them either! I am not insisting on changing anything (this was a side comment), but if you do want to try a functional test they use `JenkinsRule` and (for CLI commands) `CLICommandInvoker`.

Austry replied:
Tests now implemented with `JenkinsRule` and `CLICommandInvoker`

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,135 @@
+                executor.doStop();
+                logBuildStopped(jobName, buildName);
+            } else {
+                LOGGER.log(Level.INFO, String.format("Build %s in job %s not stopped", buildName, jobName));
```
Probably better sent to command output along with everything else.

jglick reviewed (dismissed):
> I was forced to rename

No this is incorrect and may break things. The stuff under `jenkins.CLI` is there because that is an actual class name, not a package name. Better to put your new code in the `hudson.cli` package with every other command.

(Note that if #3838 is merged, those resources get deleted.)

daniel-beck commented:
@Austry Hi, are you interested in finishing this? It seems there's review feedback that's still unaddressed.

Austry commented:
@daniel-beck Yes, I'm still interested
I will try to proceed review in a week or so

Commit:
JENKINS-11888 Remove moved files from cli package

Commit:
JENKINS-11888 Fix logging

Commit:
JENKINS-11888 Collect jobs list before stopping any builds

Commit:
JENKINS-11888 Refactor isBuilding check

Austry commented:
I fixed or answer to all review issues and rebase on fresh master branch.
@jglick can you review again? Thx!

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,125 @@
+        final List<Job> jobsToStop = new ArrayList<>();
+        for (final String jobName : names) {
+            Item item = jenkins.getItemByFullName(jobName);
+            if (item instanceof Job) {
```
Alternately, pass `Job.class` and then just do a null check.

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,125 @@
+    private void stopJobBuilds(final Job job,
+                               final String jobName) throws IOException, ServletException {
+        final Run lastBuild = job.getLastBuild();
+        final List<String> stoppedBuildsNames = new ArrayList<>();
```
As per https://github.com/jenkinsci/jenkins/pull/3686#discussion_r255682561 it would be both simpler and more responsive to throw out the code to _collect_ messages to print, and just write to `stdout` immediately upon doing something:

```
Stopped prj1 #13
Stopped prj1 #12
[…time passes…]
Stopped prj2 #99
```

That would be a somewhat different output format, of course, and if you still wanted a `No builds stopped for job …` message you would still need a flag to be set when at least one build matched.

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,125 @@
+    private void updateResultOutput(final String jobName,
+                                    final List<String> stoppedBuildNames) {
+        if (stoppedBuildNames.isEmpty()) {
+            stdout.println(String.format("No builds stopped for job '%s'", jobName));
```
By the way you can probably use `stdout.printf`.

jglick reviewed (approved):
> if #3838 is merged, those resources get deleted

For the record, it was and they were, so the `jenkins.cli` namespace is now available. Whether we actually want this one command to be in a different package from every other CLI command is another question. Best to mark it

```java
@Restricted(NoExternalUse.class)
```

to make it clear that it would be safe to move it anywhere else later.

Commit:
JENKINS-11888 Add Restricted annotation for StopBuildsCommand

Commit:
JENKINS-11888 Make cli output more responsive

daniel-beck commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        if (!isAnyBuildStopped) {
+            stdout.println("No builds stopped;");
```
Perhaps useful in the output to distinguish between a failed attempt to stop any builds, and no builds in the chosen jobs were running?

Austry replied:
Errors will be printed in output with this message.
So i think it's ok

daniel-beck commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,117 @@
+            executor.doStop();
+            stdout.println(String.format("Build %s stopped for job '%s';", buildName, jobName));
+        } else {
+            stdout.println(String.format("Build %s in job %s not stopped", buildName, jobName));
```
Why do some lines of output have semicolons at the end and others don't?

Austry replied:
Removed semicolons from all output
Think it will be more clean in that way

daniel-beck commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,117 @@
+        Executor executor = build.getExecutor();
+        if (executor != null) {
+            isAnyBuildStopped = true;
+            executor.doStop();
```
This does not gracefully handle the case when the current user does not have permission to stop the build. An exception will be thrown and not caught.

daniel-beck commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,117 @@
+        if (executor != null) {
+            isAnyBuildStopped = true;
+            executor.doStop();
+            stdout.println(String.format("Build %s stopped for job '%s';", buildName, jobName));
```
Build name is not quoted but job name is?

Commit:
JENKINS-11888 Remove single quote for job name

Commit:
JENKINS-11888 Add semicolon

Commit:
JENKINS-11888 Handle exception for executor.doStop

Commit:
JENKINS-11888 Rework test to use JenkinsRule

Commit:
JENKINS-11888 Remove semicolons from CLI output

daniel-beck commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -99,13 +99,13 @@ private void stopBuild(final Run build,
-                stdout.println(String.format("Exception occurred while trying to stop build %s for job %s;", buildName, jobName));
-                stdout.println(String.format("Exception class: %s, message: %s;", e.getClass().getSimpleName(), e.getMessage()));
+                stdout.println(String.format("Exception occurred while trying to stop build %s for job %s", buildName, jobName));
+                stdout.println(String.format("Exception class: %s, message: %s", e.getClass().getSimpleName(), e.getMessage()));
```
This should be just one line, else it will be unclear that these are related.

oleg-nenashev replied:
Agreed with @daniel-beck

Austry replied:
@oleg-nenashev It's already fixed in https://github.com/jenkinsci/jenkins/pull/3686/commits/6912362009f68b060973380568f409a3b9cee8f7

daniel-beck commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,121 @@
+            if (item instanceof Job) {
+                jobsToStop.add((Job) item);
+            } else {
+                throw new IllegalArgumentException("Invalid job name = " + jobName);
```
This could be a permissions issue, so just do…

```suggestion
                throw new IllegalArgumentException("Job not found: " + jobName);
```

daniel-beck commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,121 @@
+                               final String jobName) throws IOException, ServletException {
+        final Run lastBuild = job.getLastBuild();
+        if (lastBuild != null && lastBuild.isBuilding()) {
+            stopBuild(lastBuild, jobName);
```
Not sure about this (look at other CLI commands how they do this), but in general, output should show display names, not names.

```suggestion
            stopBuild(lastBuild, job.getFullDisplayName());
```

oleg-nenashev replied:
I would rather recommend https://javadoc.jenkins.io/hudson/model/Item.html#getFullName-- which is not localizable

daniel-beck replied:
🤷‍♂️  Should be done however other commands do it (and I'm hoping for existing consistency here).

Austry replied:
`BuildCommand` uses `job.getFullDisplayName()`
so I suggest to use `getFullDisplayName()`

daniel-beck commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,121 @@
+            try {
+                executor.doStop();
+                isAnyBuildStopped = true;
+                stdout.println(String.format("Build %s stopped for job %s", buildName, jobName));
```
Would rather quote everything than nothing. Unsure how other commands handle this.

Austry replied:
In other commands they use quotes. I'll do the same

daniel-beck reviewed (commented):
We're getting really close 😃

Commit:
JENKINS-11888 Fix tests for windows platform

Commit:
JENKINS-11888 Write message about exception in one line

Commit:
JENKINS-11888 Fix output on not founded job

Commit:
JENKINS-11888 Improve tests

Commit:
JENKINS-11888 Use fullDisplayName for job name

Commit:
JENKINS-11888 Use single quotes for build and job names

oleg-nenashev commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -0,0 +1,120 @@
+        }
+
+        for (final Job job : jobsToStop) {
+            stopJobBuilds(job);
```
If an exception happens here, would it make sense to continue and to try stopping other jobs? You could collect exceptions and then throw an aggregated one

Austry replied:
It's already work in the way you described
I added a test `shouldStopSecondJobEvenIfFirstStopFailed` to show it

oleg-nenashev reviewed (changes requested):
Thanks @Austry ! It looks almost ready to go. I would like to see the logging comment from @daniel-beck addressed, the rest are just proposals/comments

Commit:
JENKINS-11888 Cleanup exceptions throwing

Commit:
JENKINS-11888 Add test checking second job will be stopped if first stop failed

Austry commented:
I see some problem with build 
https://ci.jenkins.io/blue/organizations/jenkins/Core%2Fjenkins/detail/PR-3686/89/pipeline
Error is `Failed to install https://download.java.net/java/GA/jdk11/9/GPL/openjdk-11.0.2_windows-x64_bin.zip to C:\Jenkins\tools\hudson.model.JDK\jdk11`
Is it known build environment problem?

daniel-beck commented:
Yup, thanks. Just ignore.

batmat commented:
Retriggering build.

@Austry what is the state you consider this PR in? Do you believe you addressed comments from Oleg? Can you address specific comments, and mark each resolved as applies to clarify? Thanks

Austry commented:
> @Austry what is the state you consider this PR in? Do you believe you addressed comments from Oleg? Can you address specific comments, and mark each resolved as applies to clarify? Thanks
I believe every comment in this PR is addressed
Oleg mainly wants fix error output and in was fixed in https://github.com/jenkinsci/jenkins/commit/6912362009f68b060973380568f409a3b9cee8f7
I resolved all conversations

oleg-nenashev reviewed (approved):
I think this is ready to go. All my comments have been addressed

oleg-nenashev commented:
My plan is to merge it into the next weekly unless somebody votes against it

jglick commented on core/src/main/java/jenkins/cli/StopBuildsCommand.java:
```diff
@@ -63,63 +64,55 @@ protected int run() throws Exception {
+                isAnyBuildStopped = true;
+                stdout.println(String.format("Build '%s' stopped for job '%s'", buildName, jobName));
+            } catch (final Exception e) {
+                stdout.print(String.format("Exception occurred while trying to stop build '%s' for job '%s'. ", buildName, jobName));
```
Better to delete this `try`-`catch` block and let any exception be thrown up out of the command. The only _expected_ exception would be `AccessDeniedException`, which is better handled by `CLICommand` with a standardized exit code and everything. You do not want the command succeeding (exit code 0) if in fact an error occurred here.

Austry replied:
[Here](https://github.com/jenkinsci/jenkins/pull/3686#discussion_r277089004) was a proposal to handle `AccessDeniedException`.
And [here](https://github.com/jenkinsci/jenkins/pull/3686#discussion_r278850305) another proposal to continue stopping even if some operation failed. 
I agree that exit code 0 is wrong for situation where some build stopping failed.

Here is suggestion - handle here only 'AccessDeniedException', log every such exception and rethrow last one after attempt to stop builds for every jobs from input arguments. 
What do you think, @jglick ?

jglick commented on test/src/test/java/jenkins/cli/StopBuildsCommandTest.java:
```diff
@@ -0,0 +1,217 @@
+
+        assertThat(stdout,
+                equalTo("Exception occurred while trying to stop build '#1' for job 'jobName'. " +
+                        "Exception class: AccessDeniedException2, message: anonymous is missing the Job/Cancel permission" + LN +
```
As above, this really ought to be exit code 6 as documented in `CLICommand.main`. `CLICommandInvoker.Result` has matchers designed to be used with `assertThat`.

Austry replied:
Let's discuss [here](https://github.com/jenkinsci/jenkins/pull/3686#discussion_r281018439)

jglick commented on test/src/test/java/jenkins/cli/StopBuildsCommandTest.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    @Test
+    public void shouldStopSecondJobEvenIfFirstStopFailed() throws Exception {
```
Mm, probably not. You should not be running this command against jobs you do not have permission to abort and be expecting anything to work.

Austry replied:
Let's discuss [here](https://github.com/jenkinsci/jenkins/pull/3686#discussion_r281018439)

jglick commented on test/src/test/java/jenkins/cli/StopBuildsCommandTest.java:
```diff
@@ -0,0 +1,217 @@
+
+    private void setupSlaveWithTwoExecutors() throws hudson.model.Descriptor.FormException, IOException, URISyntaxException {
+        DumbSlave slave = new DumbSlave("slave", j.jenkins.getRootDir().getAbsolutePath(), j.createComputerLauncher(null));
+        slave.setNumExecutors(2);
```
Oddly enough, this is the default value already.

Austry replied:
You right!
`setupSlaveWithTwoExecutors()` calls unnecessary 
Will fix that

Austry replied:
Fixed in new PR https://github.com/jenkinsci/jenkins/pull/4013

jglick reviewed (commented):
The `AccessDeniedException` handling smells wrong to me but it is hardly a blocker.

oleg-nenashev commented:
Let's get it over the line then. There are indeed minor improvements, but they can be done as follow-ups if @Austry is interested. This PR has been waiting for long, and with all limitations, it is an important addition to Jenkins CLI

oleg-nenashev commented:
Thanks @Austry!
